raft/src/test/java/org/apache/kafka/raft/MockLog.java

```diff
@@ -177,14 +181,22 @@ public void appendAsFollower(Records records) {
 
     @Override
     public Records read(long startOffset, OptionalLong maxOffsetOpt) {
+        ByteBuffer buffer = ByteBuffer.allocate(512);
+
         List<LogBatch> batches = readBatches(startOffset, maxOffsetOpt);
-        if (batches.isEmpty())
-            return MemoryRecords.EMPTY;
 
-        ByteBuffer buffer = ByteBuffer.allocate(512);
         for (LogBatch batch : batches) {
             buffer = batch.writeTo(buffer);
         }
+
+        // we still need to read part of the last batch up to the max offset
```

Comment:
This is a piggy-backed fix I found from the added unit tests: the mock-log should return partial record batch as well.